test/observers.spec.js

```diff
@@ -0,0 +1,183 @@
+'use strict';
+import observers from '../src/apps/observers';
+import findFreePort from 'find-free-port';
+import nock from 'nock';
+import { Server as mockSocketServer, WebSocket } from 'mock-socket';
+import EventEmitter from 'eventemitter3';
+import { expect } from 'chai';
+import { NEW_BUNDLE } from '../src/apps/actions';
+
+const API_PREFIX = '/api/v0';
+
+class FakeStore {
+    constructor(fakeState) {
+        this.state = fakeState || {};
+        this.event = new EventEmitter();
+    }
+
+    getState() {
+        return this.state;
+    }
+
+    updateState(newState) {
+        newState = Object.assign(this.state, newState);
+        this.subscribeCB(this.state);
+    }
+
+    subscribe(cb) {
+        this.subscribeCB = cb;
+    }
+
+    dispatch(action) {
+        this.event.emit('action', action);
+    }
+}
+
+function createFakeStore(state) {
+    let fakeStore = new FakeStore(state);
+    observers(fakeStore);
+    return fakeStore;
+}
+
+describe('bundleMiddleware', () => {
+    let testPort;
+
+    beforeEach((done) => {
+        findFreePort(2000, (err, freePort) => {
+            testPort = freePort;
+            done();
+        })
+    });
+
+    afterEach(() => {
+        nock.cleanAll();
+    })
+
+    it('properly fetches bundle from null runMode', (done) => {
+        const fakeJuttle1 = {
+            program: 'emit -limit 101',
+            modules: []
+        };
+
+        nock(`http://localhost:${testPort}`)
+            .get(`${API_PREFIX}/paths/fake1.juttle`)
+            .reply(200, { bundle: fakeJuttle1 })
+            .post(`${API_PREFIX}/prepare`, {
+                bundle: fakeJuttle1,
+                inputs: {}
+            })
+            .reply(200, [])
+
+        const store = createFakeStore({
+            juttleServiceHost: `localhost:${testPort}`
+        });
+
+        store.event.once('action', (action) => {
+            expect(action).to.deep.equal({
+                type: NEW_BUNDLE,
```

Comment:
not for this pr, but `NEW_BUNDLE` is a bit misleading because `inputs` aren't typically part of a "bundle" (just program and modules) as far as the juttle-service api is concerned.